### ISSUE TITLE
Preview auto-refresh setting gets reset when browser is refreshed.

### ISSUE THREAD

**Issue by flukeout** [mozilla/thimble.mozilla.org#1633]
If you uncheck it and refresh the browser, it becomes checked again. It should stay unchecked.

![image](https://cloud.githubusercontent.com/assets/25212/22001085/b291ccc4-dbf7-11e6-8ebd-5bbbd89c567b.png)

**Comment by humphd** [comment]
Similar fix to what I describe in https://github.com/mozilla/thimble.mozilla.org/issues/1631#issuecomment-274916966

**Comment by timmoy** [comment]
Hi, I'd like to take a stab at fixing this bug.

**Comment by flukeout** [comment]
Sounds good @timmoy

**Comment by timmoy** [comment]
Hi, I'd like to take a stab at fixing this bug.

**Comment by timmoy** [comment]
After going over the code a bit, it seems that the problem is that when we refresh the browser the views/editor/nav-options.html reloads from scratch using 

 `<div class="refresh-wrapper enabled">`
        `<div title="{{ gettext("navRefreshPreviewIconTitle") }}" id="preview-pane-nav-refresh">`
          `<img src="/img/icon/refresh.svg">`
        `</div>`
       ` <div title="{{ gettext("navToggleAutoUpdateTitle") }}" class="toggle-auto-update">`
          `<span class="checkbox">`
            `<img class="checkbox-icon" src="/img/icon/checkmark-white.svg"/>`
         ` </span>`
         ` {{ gettext("navToggleAutoUpdateLabel") }}`
        `</div>`
`</div>`

The part I "enabled" in the first line is what I think is causing the issue. Whenever the user loads the page, the "enabled" class is loaded into the refresh-wrapper div again so the box becomes checked again. 

My current train of thought is that we need to somehow store the setting of whether the user wants it auto-refresh checked or not.
 
I looked through the code and found this snippet 

`<div id="navbar-login" class="{% if username %}signed-in{% endif %} nav-chunk">`

 in views/editor/userbar.html which was used to show the username if the user is signed in.

Potentially, if we edit the line I mentioned earlier into 

 `<div class="refresh-wrapper {% if auto-refresh %}enabled{% endif %}">`

and store "auto-refresh" in the user's session, it should work. 

It will also require a bit of change to the  "// Preview auto-refresh toggle" section of public/editor/scripts/editor/js/fc/bramble-ui-bridge.js as we need to change "auto-refresh" each time the user toggles. 

Any thoughts on whether this could create a conflict with other parts of thimble?

**Comment by gideonthomas** [comment]
@timmoy you're on the right track. For features that are specific to the client (front-end), we store state as preferences in `localStorage`. @humphd links to an example of how you can store/retrieve/use this value in his [comment here](https://github.com/mozilla/thimble.mozilla.org/issues/1631#issuecomment-274916966). Let's try to follow that pattern and read from `localStorage` once the page is loaded and set the auto-refresh toggle to whatever the value is in `localStorage`.

Let me know if you need any more details on this.

**Comment by timmoy** [comment]
So I managed to get the front end to display selectively on load. I added an if statement for the class itself. Not the prettiest, but it got around some of the issues I was having with my proposed fix being inside the quotes and not escaping properly.

`<div class={% if / autoUpdate %}"refresh-wrapper enabled"{% else %}" refresh-wrapper"{% endif %}>`

 I tried to then create an value and store it in a new variable in local storage using the pattern in @humphd 's comment as you described, but it seems that I'm having trouble figuring out how to replace the value whenever the user toggles the checkbox. In testing it works since if you use a variable like "username" you can have it auto checked or unchecked depending on if you are signed in or not.
I also get the logic and flow of what's happening when a user clicks:

1. User clicks the box in thimble's /views/editor/nav-options.html
2. thimble's /public/editor/scripts/editor/js/fc/bramble-ui-bridge.js sees the event, checks to see if it is enabled or disabled using hasClass("enabled") and an if statement. 
3. Based on that, it will add/remove enabled to the attribute of the tag in nav-options, and call bramble.(disable/enable)AutoUpdate() and log using analytics.event("(disable/enable)AutoUpdate")
4. (disable/enable)AutoUpdate() is found in src/bramble/client/main.js and executes a command with specified parameters and specified callback
5. The command is passed to a switch case in bramble's extensions/default/bramble/lib/RemoteCommandHandler.js and depending on the argument for "command" PostMessageTransport.setAutoUpdate(true/false) is called
6. That function is found in bramble's extensions/default/bramble/lib/PostMessageTransport.js and changes the value of _autoUpdate (and forces a reload if change to true)

But I can't seem to implement the pattern properly to store values into local storage or replicate the functionality of [humphd's fix for word wrap](https://github.com/mozilla/brackets/pull/378). I tried adapting his code to this problem, changing variable names as needed. However, the check box is now never checked, and I suspect that I'm not understanding the localstorage correctly.

However, I do know that I want to be changing the value of autoUpdate in the local storage whenever the check box is toggled so that it is saved for when the user decides to refresh. 
I think that that functionality could either in the bramble-ui-bridge.js file or in the PostMessageTransport.js files.

**Comment by humphd** [comment]
A few tips, which I hope might help:

* You have to run `npm run build` any time you change code in Bracket's `src/bramble/client.js`.  This copies the code to `dist/` which is where it actually runs from when you use Thimble.  Just in case you didn't know this, and were wondering why code changes weren't happening.
* In the bramble client API code, we manage state using this object: https://github.com/mozilla/brackets/blob/master/src/bramble/client/main.js#L184-L185.  This is implemented here: https://github.com/mozilla/brackets/blob/master/src/bramble/client/StateManager.js, and it manages reading/writing values to `localStorage`.
* You can look at how the `wordWrap` setting is stored/read here: https://github.com/mozilla/brackets/blob/master/src/bramble/client/StateManager.js#L101-L103
* When Brackets is loaded, we pass it initial state to use: https://github.com/mozilla/brackets/blob/master/src/bramble/client/main.js#L401-L418.  You'd need to include the auto refresh value here, too.  Inside Brackets, that data is received and stored in memory here: https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/main.js#L184-L194
* We then us that data in Brackets as we start the app: https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/UI.js#L87-L90.  You'd want to do something similar, but in your case, the code you need to call this: https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/RemoteCommandHandler.js#L87 
* When the `wordWrap` (or other) setting is changed in Brackets, this is where Thimble sets the state again: https://github.com/mozilla/brackets/blob/master/src/bramble/client/main.js#L298-L299

It's a complicated path, since things are happening across window boundaries (i.e., Thimble <--> Brackets) using `postMessage`, and in different repositories.

**Comment by gideonthomas** [comment]
@timmoy just to add two more things to @humphd's comment above:

- You shouldn't need to touch the `views/editor/nav-options.html` file at all. Any changes you make out there, happen at the server level before it reaches the client. We want to make changes that only touch the client. As you point out above, the change will actually be in `bramble-ui-bridge.js` on Thimble.
- This is a two repository fix tbh. There were two similar patches that landed recently, maybe they might be of help to you? The Thimble-side of the patch was: https://github.com/mozilla/thimble.mozilla.org/pull/1725 and the Brackets side was: https://github.com/mozilla/brackets/pull/584

Having changes like this span two different (large) repositories will make fixing issues like this complicated, so feel free to let us know if you need help.

**Comment by timmoy** [comment]
Thanks for the links and comments, they have been very useful!

 I am having some trouble with testing at the moment due to the getAutoUpdate() returning null (due to the getter returning null). The related code (commented stuff for later testing/debugging):
![thimble git diff 1](https://cloud.githubusercontent.com/assets/16708173/22859300/1a586928-f0a5-11e6-8e99-1f0f95655082.PNG)

I feel that the autoUpdate value is not being read correctly due to an error in how it is getting stored. So far, these are the latest stable changes I have on the bramble side following the suggestions made:
![brackets git diff 1a](https://cloud.githubusercontent.com/assets/16708173/22859353/bbf1ada2-f0a6-11e6-9949-cca0ca19c598.PNG)
![brackets git diff 1b](https://cloud.githubusercontent.com/assets/16708173/22859355/be054f0e-f0a6-11e6-83a8-62ec606ff03e.PNG)

In the RemoteEvents.js, I feel like I might need to make some changes, but the I did so a few times and ended up corrupting the brackets server data.
I tried a similar pattern to https://github.com/mozilla/brackets/pull/378/files#diff-2154d9c683f088a1bf5638db25cd5802, but with PostMessageTransport instead of PreferencesManager including the PostMessageTransport file. That didn't seem to work, and caused an error here https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/main.js#L197. I also tried to use BrambleEvents, but that didn't work either.

Am I looking in the wrong place for storing the value?

**Comment by gideonthomas** [comment]
@timmoy you would need to use `PreferencesManager` there. From what I remember, `wordWrap` is already a preference that has been incorporated in Brackets, so, the patch that you are referring to might not include some code you need to enable the preference. Since `autoUpdate` is a new preference we are creating, following [this patch](https://github.com/mozilla/brackets/pull/584/files) instead might be more helpful since it is also adding a new preference viz. `allowJavascript`. It seems like your changes to `bramble/client/StateManager.js` and `bramble/client/main.js` are correct so far. You would need to make the other changes that are suggested in the PR that I've linked.

Let me know if you need some more direction.

**Comment by timmoy** [comment]
So far, it seems to work better as it seems to be able to store a value. However, the set command seems to not work properly. 
The brackets pull request is [here](https://github.com/mozilla/brackets/pull/595).
I'll continue to fiddle around and try to find where things are going wrong.

**Comment by timmoy** [comment]
@gideonthomas your suggestion of following the other contributor's fix helped a lot to lay the groundwork of the fix. It's currently working right now after I fixed a typo from my pull requests which prevented the preference from initializing properly.

One thing to note was that when I pulled the latest changes in brackets, I found that the other fix used the register command in CommandManager ([here](https://github.com/peiying16/brackets/blob/8844b81c8141d679b96b50e8d9f373a2afdae73d/src/editor/EditorOptionHandlers.js#L104)), but in the version I was working off of, it used the registerInternal command ([here](https://github.com/timmoy/brackets/blob/c5df182bfed73dd32c48cc8665cbdc5fd55c419e/src/editor/EditorOptionHandlers.js#L107)).

This change from a global command to a global internal command actually seemed to help fix the problems I was having previously before I made the pull requests.

It's been quite a learning experience figuring out how all the code interacts. Thanks for the guidance you and @humphd have provided!
